pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,284 @@
+	defer ticker.Stop()
+
+	rw.runWithTickCh(ticker.C)
+}
+
+type revisionWatcherCh struct {
+	revisionWatcher *revisionWatcher
+	ch              chan []string
+}
+
+// RevisionBackendsManager listens to revision endpoints and keeps track of healthy
+// l4 dests which can be used to reach a revisoin
+type RevisionBackendsManager struct {
+	revisionWatchers    map[RevisionID]*revisionWatcherCh
+	revisionWatchersMux sync.RWMutex
+
+	updateCh       chan<- *RevisionDestsUpdate
+	transport      http.RoundTripper
+	logger         *zap.SugaredLogger
+	probeFrequency time.Duration
+}
+
+// NewRevisionBackendsManagerWithProbeFrequency returnes a RevisionBackendsManager that uses the supplied
+// probe frequency
+func NewRevisionBackendsManagerWithProbeFrequency(updateCh chan<- *RevisionDestsUpdate,
+	transport http.RoundTripper, endpointsInformer corev1informers.EndpointsInformer,
+	logger *zap.SugaredLogger, probeFrequency time.Duration) *RevisionBackendsManager {
+	rbm := &RevisionBackendsManager{
+		revisionWatchers: make(map[RevisionID]*revisionWatcherCh),
+		updateCh:         updateCh,
+		transport:        transport,
+		logger:           logger,
+		probeFrequency:   probeFrequency,
+	}
+
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.ChainFilterFuncs(
+			reconciler.LabelExistsFilterFunc(serving.RevisionUID),
+			// We are only interested in the private services, since that is
+			// what is populated by the actual revision backends.
+			reconciler.LabelFilterFunc(networking.ServiceTypeKey, string(networking.ServiceTypePrivate), false),
+		),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    rbm.endpointsUpdated,
+			UpdateFunc: controller.PassNew(rbm.endpointsUpdated),
+			DeleteFunc: rbm.endpointsDeleted,
+		},
+	})
+
+	return rbm
+}
+
+func NewRevisionBackendsManager(updateCh chan<- *RevisionDestsUpdate,
+	transport http.RoundTripper, endpointsInformer corev1informers.EndpointsInformer,
+	logger *zap.SugaredLogger) *RevisionBackendsManager {
+	return NewRevisionBackendsManagerWithProbeFrequency(updateCh, transport, endpointsInformer,
+		logger, probeFrequency)
+}
+
+func (rbm *RevisionBackendsManager) getOrCreateDestsCh(rev RevisionID) chan []string {
+	rbm.revisionWatchersMux.Lock()
+	defer rbm.revisionWatchersMux.Unlock()
+
+	rwCh, ok := rbm.revisionWatchers[rev]
+	if !ok {
+		destsCh := make(chan []string)
+		rw := newRevisionWatcher(rev, rbm.updateCh, destsCh, rbm.transport, rbm.logger)
+		rbm.revisionWatchers[rev] = &revisionWatcherCh{rw, destsCh}
+		go rw.run(rbm.probeFrequency)
+		return destsCh
+	}
+
+	return rwCh.ch
+}
+
+// endpointsUpdated is a handler function to be used by the Endpoints informer.
+// It updates the endpoints in the RevisionBackendsManager if the hosts changed
+//
+// This function must not be called in parallel to not induce a wrong order of events.
+func (rbm *RevisionBackendsManager) endpointsUpdated(newObj interface{}) {
```

Comment:
Can we guarantee that though? I.e. that informer won't send the same object on two different go-routines?

Reply:
This was just a copypasta comment. Not sure if this is an issue but if it is I don;t see how we could solve it from within this method - we cant guarantee anything about ordering of taking out a lock inside the method.